source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+   The convenience of using `a + b` would likely mean some places wouldn't
+   be properly bound checked, which would in turn open a DoS.
+   Instead, we provide two kind of functions for operations:
+   `bool OPNAME(ref Amount res, const Type)` and `Amount mustOPNAME(const Type)`
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Amount;
+
+/// Ditto
+public struct Amount
+{
+    import agora.common.Deserializer;
+    import agora.common.Hash;
+    import agora.common.Serializer;
+
+    import core.checkedint;
+
+    /// Number of units ('cents') per coins
+    public static immutable Amount UnitPerCoin   = Amount(10_000_000, true);
+    /// Maximum number of BOA coins that can ever be in circulation
+    public static immutable ulong  MaxCoinSupply = 500_000_000;
+    /// Maximum amount of money that can ever be in circulation
+    public static immutable Amount MaxUnitSupply =
+        Amount(UnitPerCoin.value * MaxCoinSupply, true);
+
+    /// Internal data storage
+    private ulong value;
+
+    /// Support for hashing
+    public void computeHash(scope HashDg dg) const nothrow @nogc @safe
+    {
+        hashPart(this.value, dg);
+    }
+
+    /// Support for serialization
+    public void serialize(scope SerializeDg dg) const nothrow @safe
+    {
+        serializePart(this.value, dg);
+    }
+
+    /// Support for deserialization
+    public void deserialize(scope DeserializeDg dg) nothrow @safe
+    {
+        deserializePart(this.value, dg);
+    }
+
+    /// Pretty-print this value
+    public void toString (SinkT dg) const @safe
+    {
+        import std.format;
+        formattedWrite(dg, "%d", this.value);
+    }
+
+    /// Also support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        string ret;
+        scope SinkT dg = (scope v) { ret ~= v; };
+        this.toString(dg);
+        return ret;
+    }
+
+    private alias SinkT = void delegate(scope const(char)[] v) @safe;
+
+    /// Support for Vibe.d deserialization
+    public static Amount fromString (scope const(char)[] str) pure @safe
+    {
+        import std.conv : to;
+        immutable ul = str.to!ulong;
+        if (!Amount.isInRange(ul))
+            throw new Exception("Invalid input value to Amount");
+        return Amount(ul, true);
+    }
+
+
+    nothrow pure @nogc @safe:
+
+    /***************************************************************************
+
+        Construct an instance of an `Amount`
+
+        Params:
+            units = The monetary amount this new instance represent.
+                    The unit is the smaller unit of currency (see `UnitPerCoin`)
+
+    ***************************************************************************/
+
+    public this (ulong units)
```

Comment:
Could we make a rule to move the constructor before any other methods?

Reply:
We could, but the reason it's here is because of the attributes declaration on top. Bitblob works the same.